collections/nemo_asr/nemo_asr/audio_preprocessing.py

```diff
@@ -0,0 +1,470 @@
+                                            1: AxisType(SpectrogramSignalTag),
+                                            2: AxisType(ProcessedTimeTag)}),
+
+            "processed_length": NeuralType({0: AxisType(BatchTag)})
+        }
+        return input_ports, output_ports
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+
+        self.disable_casts = (self._opt_level == Optimization.mxprO1)
+
+        self.torch_windows = {
+            'hann': torch.hann_window,
+            'hamming': torch.hamming_window,
+            'blackman': torch.blackman_window,
+            'bartlett': torch.bartlett_window,
+            'none': None,
+        }
+
+    @torch.no_grad()
+    def forward(self, input_signal, length):
+        if self.disable_casts:
+            with amp.disable_casts():
+                processed_signal = self.get_features(
+                        input_signal.to(torch.float), length)
+        else:
+            processed_signal = self.get_features(input_signal, length)
+
+        processed_length = self.get_seq_len(length)
+        return processed_signal, processed_length
+
+    def get_features(self, input_signal, length):
+        # Called by forward(). Subclasses should implement this.
+        raise NotImplementedError
+
+    def get_seq_len(self, length):
+        # Called by forward(). Subclasses should implement this.
+        raise NotImplementedError
+
+
+class AudioToSpectrogramPreprocessor(AudioPreprocessor):
+    """Preprocessor that converts wavs to spectrograms.
+    Uses torchaudio's Spectrogram class as a featurizer.
+
+    Args:
+        sample_rate (int): Sample rate of the input audio data.
+            Defaults to 16000
+        window_size (float): Size of window for fft in seconds
+            Defaults to 0.02
+        window_stride (float): Stride of window for fft in seconds
+            Defaults to 0.01
+        n_fft (int): Length of FT window. If None, it uses the smallest power
+            of 2 that is larger than n_window_size.
+            Defaults to None
+        window (str): Windowing function for fft. can be one of ['hann',
+            'hamming', 'blackman', 'bartlett']
+            Defaults to "hamming"
+        normalized (bool): Whether to normalize by magnitude after stft
+        pad (int): Two sided padding
+    """
+    def __init__(
+            self, *,
+            sample_rate=16000,
+            window_size=0.02,
+            window_stride=0.01,
+            n_fft=None,
+            window="hamming",
+            normalized=True,
+            pad=8,
+            **kwargs
+    ):
+        super().__init__(**kwargs)
+
+        self.win_length = int(sample_rate * window_size)
+        self.hop_length = int(sample_rate * window_stride)
+        self.n_fft = n_fft or 2 ** math.ceil(math.log2(self.win_length))
+
+        window_fn = self.torch_windows.get(window, None)
+        if window_fn is None:
```

Comment:
A None window type should be allowed. AKA your window is just 1.
However, an unknown window type should throw this error. This is a stuble difference depending on what you intend self.torch_windows["none"] to mean.

Reply:
It doesn't seem like `torchaudio.transforms.Spectrogram` allows us to pass in `window_fn=None`. Should we just define our own lambda for a window of 1s, in this case? Or use the torchaudio default?

Reply:
torch.stft()'s default is None. They do the same, I don't know why they changed so much.
Would mapping "none" to torch.ones work?

Reply:
Done--mapped both "none" and "null" to torch.ones (some configs use null)

Reply:
Wait a second, a null value in a yaml file converts to None. This value is different from a string "null" and different from a string "none". Let's be careful with the self.torch_windows dictionary. I would advise to remove both strings. I would allow "ones" to map to torch.ones.

We could add None: torch.ones in the self.torch_windows dict, but I am not sure about this.

Reply:
Added None to the self.torch_windows dict and removed 'null' and 'none' as per in-person conversation